pkg/api/types.go

```diff
@@ -2088,3 +2088,83 @@ type RangeAllocation struct {
 	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
 	Data []byte `json:"data"`
 }
+
+// SecurityContextConstraints governs the ability to make requests that affect the SecurityContext
+// that will be applied to a container.
+type SecurityContextConstraints struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// AllowPrivilegedContainer determines if a container can request to be run as privileged.
+	AllowPrivilegedContainer bool `json:"allowPrivilegedContainer,omitempty" description:"allow containers to run as privileged"`
+	// AllowedCapabilities is a list of capabilities that can be requested to add to the container.
+	AllowedCapabilities []Capability `json:"allowedCapabilities,omitempty" description:"capabilities that are allowed to be added"`
+	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
+	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty" description:"allow the use of the host dir volume plugin"`
+	// SELinuxContext is the strategy that will dictate what labels will be set in the SecurityContext.
+	SELinuxContext SELinuxContextStrategyOptions `json:"seLinuxContext,omitempty" description:"strategy used to generate SELinuxOptions"`
+	// RunAsUser is the strategy that will dictate what RunAsUser is used in the SecurityContext.
+	RunAsUser RunAsUserStrategyOptions `json:"runAsUser,omitempty" description:"strategy used to generate RunAsUser"`
+
+	// The users who have permissions to use this security context constraints
+	Users []string `json:"users,omitempty" description:"users allowed to use this SecurityContextConstraints"`
+	// The groups that have permission to use this security context constraints
+	Groups []string `json:"groups,omitempty" description:"groups allowed to use this SecurityContextConstraints"`
+}
+
+// SELinuxContextStrategyOptions defines the strategy type and any options used to create the strategy.
+type SELinuxContextStrategyOptions struct {
+	// Type is the strategy that will dictate what SELinux context is used in the SecurityContext.
+	Type SELinuxContextStrategyType `json:"type,omitempty" description:"strategy used to generate the SELinux context"`
```

Comment:
Omitempty doesn't really mean optional.  It means, don't clutter up the JSON output with this key if it's not specified.

----- Original Message -----
> > +	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty"
> > description:"allow the use of the host dir volume plugin"`
> > +	// SELinuxContext is the strategy that will dictate what labels will be
> > set in the SecurityContext.
> > +	SELinuxContext SELinuxContextStrategyOptions
> > `json:"seLinuxContext,omitempty" description:"strategy used to generate
> > SELinuxOptions"`
> > +	// RunAsUser is the strategy that will dictate what RunAsUser is used in
> > the SecurityContext.
> > +	RunAsUser RunAsUserStrategyOptions `json:"runAsUser,omitempty"
> > description:"strategy used to generate RunAsUser"`
> > +
> > +	// The users who have permissions to use this security context
> > constraints
> > +	Users []string `json:"users,omitempty" description:"users allowed to use
> > this SecurityContextConstraints"`
> > +	// The groups that have permission to use this security context
> > constraints
> > +	Groups []string `json:"groups,omitempty" description:"groups allowed to
> > use this SecurityContextConstraints"`
> > +}
> > +
> > +// SELinuxContextStrategyOptions defines the strategy type and any options
> > used to create the strategy.
> > +type SELinuxContextStrategyOptions struct {
> > +	// Type is the strategy that will dictate what SELinux context is used in
> > the SecurityContext.
> > +	Type SELinuxContextStrategyType `json:"type,omitempty"
> > description:"strategy used to generate the SELinux context"`
> 
> Is this really optional? `omitempty` means optional.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7893/files#r31054623